### PR TITLE
Refactors update thread

### PR DESCRIPTION
**Pull request by Rotorsoft**
## Link to Issue
Closes: #8932

## Description of Changes
- Refactors old endpoint
- Moves tests to lifecycle

## Test Plan
- Should be able to update thread's attributes 
- @timolegros discordbot integration is failing when getting post owner in listener

## Deployment Plan
- Deploy discord-bot

### CODE REVIEW

File: libs/model/src/thread/UpdateThread.command.ts [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+        collaboratorsPatch.remove.length > 0
+      ) {
+        const contestManagers = await models.sequelize.query(
+          `
+SELECT cm.contest_address FROM "Threads" t
+JOIN "ContestTopics" ct on ct.topic_id = t.topic_id
+JOIN "ContestManagers" cm on cm.contest_address = ct.contest_address
+WHERE t.id = :thread_id
+`,
+          {
```
Comment (rbennettcw) [inline libs/model/src/thread/UpdateThread.command.ts:214]: Query looks fine. There's a helper called `getThreadContestManagers` but it only gets the *active* contest managers, while in this context, we want to check if the thread was ever part of a contest.

**Review by rbennettcw — APPROVED** [review]
Tested and works.

**Review by timolegros — COMMENTED** [review]
Some nits/comments that can potentially be handled in follow-up tickets. I extensively tested all update functionality (admin and non-admin) and found no issues. Good stuff!

Happy to approve once my questions are answered and/or we have follow-up tickets.

File: libs/model/src/middleware/authorization.ts [hunk 2]
```diff
@@ -128,6 +129,11 @@ async function authorizeAddress(
     } else {
       auth.thread = await models.Thread.findOne({
         where: { id: auth.thread_id },
+        include: {
```
[the remark this answers is not in the document]
Comment (Rotorsoft) [reply libs/model/src/middleware/authorization.ts:132]: done

File: libs/model/src/middleware/authorization.ts [hunk 3]
```diff
@@ -232,25 +238,52 @@ export const isSuperAdmin: AuthHandler = async (ctx) => {
...
-    if (ctx.actor.user.isAdmin) return;
-    const auth = await authorizeAddress(ctx, roles);
+    const isAdmin = ctx.actor.user.isAdmin;
+
+    const auth = await buildAuth(
+      ctx,
+      isAdmin ? ['admin', 'moderator', 'member'] : roles,
+    );
+
+    if (isAdmin) return;
```
[the remark this answers is not in the document]
Comment (Rotorsoft) [reply libs/model/src/middleware/authorization.ts:265]: we want to build the auth context for super admins too... Malik/Raymond/Jake found this bug yesterday, not allowing super admins to create threads b/c the mustBeAuthenticated guard was failing

File: libs/model/src/middleware/authorization.ts [hunk 4]
```diff
@@ -232,25 +238,52 @@ export const isSuperAdmin: AuthHandler = async (ctx) => {
...
+
     if (auth.address!.is_banned) throw new BannedActor(ctx.actor);
-    if (auth.author_address_id && auth.address!.id === auth.author_address_id)
-      return; // author
-    if (action && auth.address!.role === 'member')
-      await isTopicMember(ctx.actor, auth, action);
+
+    auth.is_author = auth.address!.id === auth.author_address_id;
+    if (auth.is_author) return;
```
[the remark this answers is not in the document]
Comment (Rotorsoft) [reply libs/model/src/middleware/authorization.ts:270]: Agree, let's ask @dillchen @zakhap @jnaviask about this... the current auth waterfall considers "content authors" a higher authorization level than "group permissions"

File: libs/model/src/thread/UpdateThread.command.ts [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+      // check if patch violates contest locks
+      if (
+        Object.keys(content).length > 0 ||
+        ownerPatch.topic_id ||
+        collaboratorsPatch.add.length > 0 ||
+        collaboratorsPatch.remove.length > 0
+      ) {
+        const contestManagers = await models.sequelize.query(
+          `
+SELECT cm.contest_address FROM "Threads" t
```
[the remark this answers is not in the document]
Comment (Rotorsoft) [reply libs/model/src/thread/UpdateThread.command.ts:209]: Good point... I was just copy/pasting the old endpoint but I'm refactoring to sequelize in next commit

File: libs/model/src/thread/UpdateThread.command.ts [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+                community_id: thread.community_id,
+              }));
+          }
+        }
+      });
+      // == end of transaction boundary ==
+
+      // TODO: should we make a query out of this, or do we have one already?
+      return (
+        await models.Thread.findOne({
```
[the remark this answers is not in the document]
Comment (Rotorsoft) [reply libs/model/src/thread/UpdateThread.command.ts:298]: But I think these are expected by the UI (@mzparacha to confirm).... we should also have a single query to return a thread by id that can be reused here (see comment)... let's keep it like this and create a new refactoring ticket to resolved after Malik's comments

File: packages/commonwealth/client/scripts/state/api/threads/editThread.ts [hunk 7]
```diff
@@ -123,9 +118,10 @@ const useEditThreadMutation = ({
   const utils = trpc.useUtils();
   const { checkForSessionKeyRevalidationErrors } = useAuthModalStore();
 
-  return useMutation({
-    mutationFn: editThread,
-    onSuccess: async (updatedThread) => {
+  return trpc.thread.updateThread.useMutation({
+    onSuccess: (updated) => {
+      // @ts-expect-error StrictNullChecks
```
[the remark this answers is not in the document]
Comment (Rotorsoft) [reply packages/commonwealth/client/scripts/state/api/threads/editThread.ts:123]: We have these in a few places in the UI. We should probably have a cleanup bucket dedicated to reconciling new schemas with the UI and remove all these model adapters, but this is going to be a long project. cc @mzparacha

File: libs/model/src/thread/UpdateThread.command.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Rotorsoft) [reply libs/model/src/thread/UpdateThread.command.ts]: @rbennettcw can you double check the new Sequelize query is OK?

File: libs/model/src/middleware/authorization.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (dillchen) [reply libs/model/src/middleware/authorization.ts:276]: Good question, we should do the reverse

1. Group Permission
2. Author Setting

So if group permission - is "no" for update poll
And Auth setting - is yes

We should default group permission

File: libs/model/src/thread/UpdateThread.command.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (mzparacha) [reply libs/model/src/thread/UpdateThread.command.ts:292]: Yes, comments/reactions and topics on the thread object are used by the UI in many instances, ex: see [here](https://github.com/hicommonwealth/commonwealth/blob/master/packages/commonwealth/server/controllers/server_threads_methods/get_active_threads.ts#L121) and [here](https://github.com/hicommonwealth/commonwealth/blob/master/packages/commonwealth/client/scripts/views/pages/discussions/ThreadCard/ThreadOptions/ReactionButton/ReactionButton.tsx#L52) to name a few.

File: libs/model/src/middleware/authorization.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (Rotorsoft) [reply libs/model/src/middleware/authorization.ts:281]: done... pls review final waterfall to confirm

[the remark this answers is not in the document]
Comment (dillchen) [reply libs/model/src/middleware/authorization.ts]: it LGTM

**Review by dillchen — APPROVED** [review]

File: libs/model/src/thread/UpdateThread.command.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (timolegros) [reply libs/model/src/thread/UpdateThread.command.ts:292]: @mzparacha My point is not that we don't need this data in the UI, my point is that this data already exists on the client before we update the thread. It's not like by updating the thread we suddenly have new comments, reactions, and topics.

Is this kind of caching something that we can do down the road @mzparacha? If so can you create a ticket for this?

**Review by timolegros — APPROVED** [review]

File: packages/commonwealth/client/scripts/state/api/threads/editThread.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (mzparacha) [reply packages/commonwealth/client/scripts/state/api/threads/editThread.ts:123]: I think @burtonator had a plan to remove those, but yeah I agree we should remove `StrictNullChecks` whenever possible.

File: libs/model/src/thread/UpdateThread.command.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (mzparacha) [reply libs/model/src/thread/UpdateThread.command.ts:292]: Yeah, the data would already exist on FE if any update endpoint is called. 

Previously, There was a lot of cache complexity in frontend to handle different response formats of the same entity from different routes specially with threads -- consistent response formats were added to some endpoints to solve some of those issues during https://github.com/hicommonwealth/commonwealth/pull/7958.

To confirm your point, do you mean we only return the updated fields? 

If yes, then I suggest we follow this pattern consistently with every PATCH/PUT route and modify the frontend accordingly. 

cc: @timolegros @Rotorsoft 

cc: @timolegros @Rotorsoft

[the remark this answers is not in the document]
Comment (Rotorsoft) [reply libs/model/src/thread/UpdateThread.command.ts:292]: Thanks @mzparacha ... do you already have a ticket/bucket to refactor the UI according to ⬆️ ?

**Pull request merged**